lms/djangoapps/bookmarks/models.py

```diff
@@ -36,23 +49,73 @@ def create(cls, bookmark_data):
         Raises:
             ItemNotFoundError: If no block exists for the usage_key.
         """
-        usage_key = bookmark_data.pop('usage_key')
+        data = copy.deepcopy(data)
+
+        usage_key = data.pop('usage_key')
         usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
 
         block = modulestore().get_item(usage_key)
 
-        bookmark_data['course_key'] = usage_key.course_key
-        bookmark_data['display_name'] = block.display_name
-        bookmark_data['path'] = cls.get_path(block)
-        user = bookmark_data.pop('user')
+        xblock_cache = XBlockCache.create({
+            'usage_key': usage_key,
+            'display_name': block.display_name,
+        })
 
-        bookmark, __ = cls.objects.get_or_create(usage_key=usage_key, user=user, defaults=bookmark_data)
+        data['course_key'] = usage_key.course_key
+        data['xblock_cache'] = xblock_cache
+
+        if xblock_cache.paths and len(xblock_cache.paths) == 1:
+            # If there is only one path to the block, use it.
+            data['path'] = xblock_cache.paths[0]
+        else:
+            # In case of multiple paths, in the future get_path_data()
+            # can be updated to look at which path the user has access to.
+            data['path'] = Bookmark.get_path_data(usage_key)
+
+        user = data.pop('user')
+
+        bookmark, __ = cls.objects.get_or_create(usage_key=usage_key, user=user, defaults=data)
         return bookmark
 
+    @property
+    def display_name(self):
+        """
+        Return the display_name from self.xblock_cache.
+
+        Returns:
+            String.
+        """
+        return self.xblock_cache.display_name
+
+    @property
+    def updated_path(self):
+        """
+        Return the latest path to this block after checking self.xblock_cache.
+
+        Returns:
+            List of dicts.
+        """
+        if self.modified < self.xblock_cache.modified:
+
+            if self.xblock_cache.paths and len(self.xblock_cache.paths) == 1:
```

Comment:
@cpennington This check if for both. The reason I am not picking the first one here is that it may be different from what `jump_to_id` will take the user to, and that will be consistent with what `path_to_location` returns.